File: Data Structures & Algorithms/surrounded-regions/submission-4.cs
public class Solution {
    public void Solve(char[][] board) {
        int ROW = board.Length;
        int COL = board[0].Length;
        for(int i=0;i<ROW;i++) {
            for(int j=0;j<COL;j++){
                if(i==0||j==0||i==ROW-1||j==COL-1){
                    if(board[i][j]=='O') dfs(i,j,board);
                }
            }
        }
        for(int i=0;i<ROW-1;i++){
            for(int j=0;j<COL-1;j++){
                if(board[i][j]=='O') board[i][j] = 'X';
            }
        }
        for(int i=0;i<ROW;i++) {
            for(int j=0;j<COL;j++){
                if(board[i][j]=='T') board[i][j] = 'O';
            }
        }
    }

    public void dfs(int r, int c, char[][] board){
        if(r<0||c<0||r==board.Length||c==board[0].Length||board[r][c]!='O'){
            return;
        }
        board[r][c] = 'T';
        dfs(r+1,c,board);
        dfs(r-1,c,board);
        dfs(r,c+1,board);
        dfs(r,c-1,board);
    }
}

File: Data Structures & Algorithms/search-for-word/submission-4.java
class Solution {
    public boolean exist(char[][] board, String word) {
       boolean visit[][] = new boolean[board.length][board[0].length];
       for(int i=0;i<board.length;i++){
        for(int j=0;j<board[0].length;j++){
            if(word.charAt(0)==board[i][j] && dfs(i,j,0,board,word,visit)) return true;
        }
       }
       return false; 
    }
    public boolean dfs(int r, int c, int k, char[][] board, String word, boolean visit[][]){
        if(k==word.length()) return true;
        if(r<0||c<0||r==board.length||c==board[0].length||visit[r][c]||board[r][c]!=word.charAt(k)) return false;
        visit[r][c] = true;
        boolean res = dfs(r+1, c, k+1, board, word, visit)||dfs(r-1, c, k+1, board, word, visit)||dfs(r, c+1, k+1, board, word, visit)||dfs(r, c-1, k+1, board, word, visit);
        visit[r][c] = false;
        return res;
    }
}
